extensions/replication/queueProcessor/QueueProcessor.js

```diff
@@ -799,8 +799,7 @@ class QueueProcessor {
             zookeeper: this.zkConfig,
             topic: this.repConfig.topic,
             groupId: this.repConfig.queueProcessor.groupId,
-            concurrency: 1, // replication has to process entries in
-                            // order, so one at a time
+            concurrency: 10, // versioning can support out of order updates
```

Comment:
You will also need to spawn separate contexts for every parallel processing "thread". Look for `// per-request state variables` comment in the constructor, normally the member variables following are to be duplicated per replicated entry. The cleanest way would likely be to create a new class containing those variables and the methods around processing a single entry, then the QueueProcessor class would become the orchestrator only.

Also please double-check that we're not getting into trouble with version-specific operations (I can think of put-object-tagging, or put-object-acl, which if done out-of-order may lead to inconsistent replication).

Otherwise once we sort out details I'm all for this improvement, as it should allow much fewer partitions and processes running for the same throughput performance (so less memory consumed, CPU overhead etc.) and simpler config since we will care less about adjusting the number of partitions to the workload.

Reply:
Ah yes, I now remember the conversation we had earlier. Thanks for pointing it out. We should isolate the state for each entry while sharing the same client instances (may be).